CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+fail-overs.
+If the target Failover Coordinator is still unaware of the new topology it might cause the shards to ping-pong a bit between the Failover Coordinators, so a node should contain logic to reduce
+the frequency of heartbeats after a moved response, but even if it doesn't this situation will self-resolve without
+causing any damage.
+
+Worth noting that this flow is the main reason why a Topology Director **must** ensure no concurrent updates to the same shard:
+- Shard X comprised of 3 nodes X1,X2,X3 belong to Failover Coordinator A
+- X is moved to Failover Coordinator B
+- All entities other than X3 get the updated topology
+- X1 sends a heartbeat as primary to Failover Coordinator **B**
+- X is moved back to Failover Coordinator A, but Failover Coordinator B is delayed in receiving the updated topology
+- X3 sends a heartbeat as primary to Failover Coordinator **A**
+While this situation will self-resolve when Failover Coordinator B receives the updated topology, Flotilla aims to guarantee a strongly 
+consistent topology - so it needs to prevent concurrent updates.
+
+![Move Shard](images/move_shards.jpg?raw=true "Move shard between Failover Coordinators")
+
+### Failover
+
+Every K milliseconds, every node sends a heartbeat to the Failover Coordinator which owns it. This heartbeat includes:
+- Replication offset (for replicas)
+- Last known cluster topology epoch
+- Last known Shard topology epoch
+- Role (primary/replica)
+
+The Failover Coordinator can respond with the following:
+- ACK: The node topology is up-to-date. The heartbeat data is saved in the Failover Coordinator along with the current tick.
+- NACK: The node Topology is out-of-date
+- MOVED: The node does not belong to the Failover Coordinator. This response includes the details of the relevant Failover Coordinator.
+
+The response includes:
+- The heartbeat results from all the nodes in the same shard (and their 'tick' value)
+- If needed, the updated cluster topology.
+
+If a replica node detects that its primary has not sent a heartbeat for more than N ticks it checks if it is the most
+up-to-date replica using the reported replication offset. If it is, then it sends a heartbeat request with an updated
+role and an incremented shard epoch. If more than one node has the most recent offset, then the node ID is used
+as tiebreaker. When the Failover Coordinator notes the epoch change it modifies the heartbeats of all other nodes in the shard to
+the replica role.
+Note that if multiple nodes attempt to promote themselves only one can succeed (The first one to be processed by the Failover Coordinator)
+All the rest will have their heartbeat rejected with a NACK.
+A planned failover works very similarly to the current CLUSTER FAILOVER command and can also support the FORCE flag (The
+TAKEOVER flag _is not_ supported as it violated the topology consistency by definition). Once the replica sees its
```

Comment:
I don't see why takeover should be "unsupported". If the FC is down for some reason, and the cluster is stuck, using a failover 'takeover' seems appropriate. When the FC comes back up, it should re-learn about the new state and either force it back *or* accept the new state. Either that, or we take the hard stance that mastership is 'immutable' when the FC is down. Note, I think it's reasonable to say 'topology' is immutable when the TD is down.

Another unrelated comment, I feel like we should actually have a `CLUSTER STEPDOWN` API, to stepdown to any replica who is eligible to takeover. Also, I just realized there is a lack of parity between `FAILOVER` and `CLUSTER FAILOVER`. One of them you execute on a primary to failover to a replica, the other you execute on a replica to takeover from the primary. I might open a separate issue for this.

Reply:
The design says that mastership is immutable if the FC is down.
TAKEOVER allows one to potentially have two primaries at the same time, with no good way to automatically resolve the conflict without data loss once the control plane is back up.
Consider the following:
- Nodes A,B,C in Shard X, A is primary
- FC goes down
- Admin runs TAKEOVER to promote B, but neither C nor A have any way of learning about this, and A still considers itself the primary
- FC comes back up
- C is able to re-establish the connection first, and promotes itself
- A & B heartbeat and establish themselves as replicas
(I can think of another reconnection order which would also break)

RE STEPDOWN, I don't think it's really specific to Flotille, If there's a strong enough need and it's implemented in the current clustering scheme we can easily implement it here as well.
Might be worth noting that removng a primary node in Flotilla does something similar (the removed primary will drain all client connections and stop sending heartbeat)

Reply:
Yes, availability over data is the tradeoff, which many applications are happy to make. There is a way, which is just pick one, specifically whichever one the FC says when it comes back up.

> Might be worth noting that removng a primary node in Flotilla does something similar (the removed primary will drain all client connections and stop sending heartbeat)

A primary in what context? In the raft groups?

Reply:
I meant in the context of each shard, though it's a similar thing with Raft (the leader demotes itself before removing from the quorum).

Reply:
> I don't see why takeover should be "unsupported". If the FC is down for some reason, and the cluster is stuck, using a failover 'takeover' seems appropriate. When the FC comes back up, it should re-learn about the new state and either force it back or accept the new state. Either that, or we take the hard stance that mastership is 'immutable' when the FC is down. Note, I think it's reasonable to say 'topology' is immutable when the TD is down.

It will create another path for Primary/replica CP flow. Currently, the FC is the process owner. If the TD can override the Primary role, it also means that the nodes can receive the roles and states from the TD directly.

I wonder if having a better/easy way to fix the FC is not good enough to unstuck the user.

Reply:
> I wonder if having a better/easy way to fix the FC is not good enough to unstuck the user.

This might actually be the correct answer. I've been through enough weird issues at AWS that there is going to be a long tail of issues around this new clustering system, and we should make sure we have enough tooling to support those unlucky enough to run into them.

Reply:
I've added a comment in the "troubleshooting" placeholder that we need to address this specific issue